readme.txt
=== Decorator - WooCommerce Email Customizer ===
Contributors: RightPress
Tags: woocommerce email customizer, woocommerce, customize, customizer, customise, customiser, email, preview, style, template, decorator
Donate link: https://codecanyon.net/user/rightpress/portfolio
Requires at least: 4.4
Tested up to: 4.7
Stable tag: 1.0.1
License: GPLv3 or later
License URI: http://www.gnu.org/licenses/gpl-3.0.html

Use native WordPress Customizer to make WooCommerce emails match your brand. Easy to use WooCommerce Email Customizer.

== Description ==
Love the ease with which you can make changes to your WordPress site with the built-in Customizer tool? Don't like the standard look of the WooCommerce emails too much? This WooCommerce email customizer extension is exactly what you need!

This plugin does only one thing but does it well - it sets up a separate instance of the Customizer tool and allows you to make changes to WooCommerce emails with live preview.

Customize your WooCommerce emails now and stand out from the crowd!

Created at [RightPress](https://codecanyon.net/user/rightpress/portfolio?ref=RightPress) by people who know what they are doing.

<small>Other plugins by RightPress: [WooCommerce Email Center](https://codecanyon.net/item/woocommerce-email-center/13907681?ref=RightPress), [WooShip - WooCommerce Shipping Plugin](https://codecanyon.net/item/wooship-woocommerce-shipping-plugin/13593884?ref=RightPress), [WooCommerce Custom Fields & Product Add-ons](https://codecanyon.net/item/woocommerce-custom-fields-product-addons/11332742?ref=RightPress), [Subscriptio - WooCommerce Subscriptions](http://codecanyon.net/item/subscriptio-woocommerce-subscriptions/8754068?ref=RightPress), [WooCommerce Membership](http://codecanyon.net/item/woocommerce-membership/8746370?ref=RightPress), [WooCommerce Quote](http://codecanyon.net/item/woocommerce-quote/7411885?ref=RightPress), [WooCommerce Dynamic Pricing & Discounts](http://codecanyon.net/item/woocommerce-dynamic-pricing-discounts/7119279?ref=RightPress), [Chimpy - MailChimp WordPress Plugin](http://codecanyon.net/item/chimpy-mailchimp-wordpress-plugin/6290216?ref=RightPress), [WooChimp - WooCommerce MailChimp Integration](http://codecanyon.net/item/woochimp-woocommerce-mailchimp-integration/6044286?ref=RightPress), [WooCommerce PDF Invoice](http://codecanyon.net/item/woocommerce-pdf-invoice/5951088?ref=RightPress)</small>

== Screenshots ==

1. WooCommerce Email Settings Page
2. Standard WooCommerce Email Styling
3. Custom WooCommerce Email Styling

== Changelog ==
= Version 1.0, 12 December 2016 =
* Initial release


changelog
Version 1.0.3, 8 May 2017
------------------------------------------------------------------------------------
* Fix - Warning related to second parameter of date()

Version 1.0.2, 4 April 2017
------------------------------------------------------------------------------------
* Fix - Styles are not always applied to order items table

Version 1.0.1, 25 February 2017
------------------------------------------------------------------------------------
* Tweak - Improved compatibility with WooCommerce 2.7

Version 1.0, 12 December 2016
------------------------------------------------------------------------------------
* Initial release


includes/classes/rp-decorator-preview.class.php
<?php

// Exit if accessed directly
if (!defined('ABSPATH')) {
    exit;
}

/**
 * Preview for WordPress Customizer
 *
 * @class RP_Decorator_Preview
 * @package Decorator
 * @author RightPress
 */
if (!class_exists('RP_Decorator_Preview')) {

class RP_Decorator_Preview
{
    // WooCommerce email classes
    public static $email_classes = array(
        'WC_Email_New_Order'                    => 'processing',
        'WC_Email_Cancelled_Order'              => 'cancelled',
        'WC_Email_Failed_Order'                 => 'failed',
        'WC_Email_Customer_On_Hold_Order'       => 'on-hold',
        'WC_Email_Customer_Processing_Order'    => 'processing',
        'WC_Email_Customer_Completed_Order'     => 'completed',
        'WC_Email_Customer_Refunded_Order'      => 'refunded',
        'WC_Email_Customer_Invoice'             => 'processing',
        'WC_Email_Customer_Note'                => 'processing',
        'WC_Email_Customer_Reset_Password'      => null,
        'WC_Email_Customer_New_Account'         => null,
    );

    // Singleton instance
    private static $instance = false;

    /**
     * Singleton control
     */
    public static function get_instance()
    {
        if (!self::$instance) {
            self::$instance = new self();
        }
        return self::$instance;
    }

    /**
     * Class constructor
     *
     * @access public
     * @return void
     */
    public function __construct()
    {
        // Set up preview
        add_action('parse_request', array($this, 'set_up_preview'));
    }

    /**
     * Set up preview
     *
     * @access public
     * @return void
     */
    public function set_up_preview()
    {
        // Make sure this is own preview request
        if (!RP_Decorator::is_own_preview_request()) {
            return;
        }

        // Load main view
        include RP_DECORATOR_PLUGIN_PATH . 'includes/views/preview.php';

        // Do not load any further elements
        exit;
    }

    /**
     * Print preview email
     *
     * @access public
     * @return void
     */
    public static function print_preview_email()
    {
        // Load WooCommerce emails
        $wc_emails = WC_Emails::instance();
        $emails = $wc_emails->get_emails();

        // Reference email
        if (isset($emails['WC_Email_Customer_Processing_Order']) && is_object($emails['WC_Email_Customer_Processing_Order'])) {
            $email = $emails['WC_Email_Customer_Processing_Order'];
            $order_status = 'processing';
        }
        else {
            // Note: currently this is only used as a fallback
            foreach (RP_Decorator_Preview::$email_classes as $email_class => $order_status) {
                if (isset($emails[$email_class]) && is_object($emails[$email_class])) {
                    $email = $emails[$email_class];
                    break;
                }
            }
        }

// Note: need to set specific properties for specific emails when implementing full support for different emails

// NEW ACCOUNT
// $this->object             = new WP_User( $user_id );
// $this->user_pass          = $user_pass;
// $this->user_login         = stripslashes( $this->object->user_login );
// $this->user_email         = stripslashes( $this->object->user_email );
// $this->recipient          = $this->user_email;
// $this->password_generated = $password_generated;

// CUSTOMER NOTE
// $this->customer_note           = $customer_note;

// REFUNDED ORDER
// $this->partial_refund = $partial_refund;
// $this->set_email_strings( $partial_refund );    <-- run this after setting $partial_refund

// RESET PASSWORD
// $this->object     = get_user_by( 'login', $user_login );
// $this->user_login = $user_login;
// $this->reset_key  = $reset_key;
// $this->user_email = stripslashes( $this->object->user_email );
// $this->recipient  = $this->user_email;

        // Order object
        $email->object = RP_Decorator_Preview::get_wc_order_for_preview($order_status);

        // Macros
        $email->find['order-date']      = '{order_date}';
        $email->find['order-number']    = '{order_number}';
        $email->replace['order-date']   = date_i18n(wc_date_format(), strtotime($email->object->order_date));
        $email->replace['order-number'] = $email->object->get_order_number();

        // Other properties
        $email->recipient = $email->object->billing_email;

        // Get email content and apply styles
        $content = $email->get_content_html();
        $content = $email->style_inline($content);

        // Print email content
        echo $content;

        // Print live preview scripts in footer
        add_action('wp_footer', array('RP_Decorator_Preview', 'print_live_preview_scripts'), 99);
    }

    /**
     * Get WooCommerce order for preview
     *
     * @access public
     * @param string $order_status
     * @return object
     */
    public static function get_wc_order_for_preview($order_status = null)
    {
        // Get last WooCommerce order
        $last_order_id = get_posts(array(
            'numberposts'   => 1,
            'orderby'       => 'ID',
            'post_type'     => 'shop_order',
            'post_status'   => array('wc-pending', 'wc-processing', 'wc-on-hold', 'wc-completed'),
            'fields'        => 'ids',
        ));

        // Use real order
        if (!empty($last_order_id)) {
            $last_order_id = array_pop($last_order_id);
            return wc_get_order($last_order_id);
        }
        // Use mockup order (WC 2.7+)
        else if (RP_Decorator::wc_version_gte('2.7')) {

            // Instantiate order object
            $order = new WC_Order();

            // Other order properties
            $order->set_props(array(
                'id'                    => 1,
                'status'                => ($order_status === null ? 'processing' : $order_status),
                'billing_first_name'    => 'Sherlock',
                'billing_last_name'     => 'Holmes',
                'billing_company'       => 'Detectives Ltd.',
                'billing_address_1'     => '221B Baker Street',
                'billing_city'          => 'London',
                'billing_postcode'      => 'NW1 6XE',
                'billing_country'       => 'GB',
                'billing_email'         => 'sherlock@holmes.co.uk',
                'billing_phone'         => '02079304832',
                'date_created'          => date('Y-m-d H:i:s'),
                'total'                 => 24.90,
            ));

            // Item #1
            $order_item = new WC_Order_Item_Product();
            $order_item->set_props(array(
                'name'      => 'A Study in Scarlet',
                'subtotal'  => '9.95',
            ));
            $order->add_item($order_item);

            // Item #2
            $order_item = new WC_Order_Item_Product();
            $order_item->set_props(array(
                'name'      => 'The Hound of the Baskervilles',
                'subtotal'  => '14.95',
            ));
            $order->add_item($order_item);

            // Return mockup order
            return $order;
        }
        // Use mockup order (pre WC 2.7)
        else {

            // Include mockup order class
            if (!class_exists('RP_Decorator_Mockup_Order')) {
                include RP_DECORATOR_PLUGIN_PATH . 'includes/classes/lazy/rp-decorator-mockup-order.class.php';
            }

            // Instantiate order object
            $order = new RP_Decorator_Mockup_Order();

            // Set order status
            if ($order_status) {
                $order->status = $order_status;
            }

            // Return mockup order
            return $order;
        }
    }

    /**
     * Print live preview scripts
     *
     * @access public
     * @return void
     */
    public static function print_live_preview_scripts()
    {
        // Open container
        $scripts = '<script type="text/javascript">jQuery(document).ready(function() {';

        // Font family mapping
        $scripts .= 'var font_family_mapping = ' . json_encode(RP_Decorator_Settings::$font_family_mapping) . ';';

        // Function to handle special cases
        $scripts .= "function prepare(value, key, selector) {
            if (key === 'border_radius' && selector === '#template_header') {
                value = value.replace('!important', '').trim();
                value = value + ' ' + value + ' 0 0 !important';
            }
            else if (key === 'email_padding' && selector === '#wrapper') {
                value = value + ' 0 ' + value + ' 0';
            }
            else if (key === 'footer_padding' && selector === '#template_footer #credit') {
                value = '0 ' + value + ' ' + value + ' ' + value;
            }
            else if (key === 'footer_content_text' && value !== '') {
                value = '<p>' + value + '</p>';
            }
            else if (key === 'shadow') {
                value = '0 ' + (value > 0 ? 1 : 0) + 'px ' + (value * 4) + 'px ' + value + 'px rgba(0,0,0,0.1) !important';
            }
            else if (key.match(/font_family$/)) {
                value = typeof font_family_mapping[value] !== 'undefined' ? font_family_mapping[value] : value;
            }
            else if (key === 'header_image') {
                value = '<p style=\"margin-top:0;\"><img src=\"' + value + '\" style=\"border: none; display: inline; font-weight: bold; height: auto; line-height: 100%; outline: none; text-decoration: none; text-transform: capitalize;\" /></p>';
            }
            return value;
        }";

        // Get CSS suffixes
        $scripts .= 'var suffixes = ' . json_encode(RP_Decorator_Customizer::get_css_suffix()) . ';';

        // Iterate over settings
        foreach (RP_Decorator_Settings::get_settings() as $setting_key => $setting) {

            // No live method
            if (!isset($setting['live_method'])) {
                continue;
            }

            // Iterate over selectors
            if (in_array($setting['live_method'], array('css', 'property')) && !empty($setting['selectors'])) {
                foreach ($setting['selectors'] as $selector => $properties) {

                    // Iterate over properties
                    foreach ($properties as $property) {

                        // CSS value change
                        if (!isset($setting['live_method']) || $setting['live_method'] === 'css') {
                            $scripts .= "wp.customize('rp_decorator[$setting_key]', function(value) {
                                value.bind(function(newval) {
                                    newval = newval + (typeof suffixes['$setting_key'] !== 'undefined' ? suffixes['$setting_key'] : '');
                                    newval = prepare(newval, '$setting_key', '$selector');
                                    jQuery('$selector').css('$property', '').attr('style', function(i, s) { return (s||'') + '$property: ' + newval + ';' });
                                });
                            });";
                        }

                        // DOM object property
                        if ($setting['live_method'] === 'property') {
                            $scripts .= "wp.customize('rp_decorator[$setting_key]', function(value) {
                                value.bind(function(newval) {
                                    newval = newval + (typeof suffixes['$setting_key'] !== 'undefined' ? suffixes['$setting_key'] : '');
                                    newval = prepare(newval, '$setting_key', '$selector');
                                    jQuery('$selector').prop('$property', newval);
                                });
                            });";
                        }
                    }
                }
            }

            // HTML Replace
            if ($setting['live_method'] === 'replace' && !empty($setting['selectors'])) {
                foreach ($setting['selectors'] as $selector) {
                    $original = json_encode($setting['original']);
                    $scripts .= "wp.customize('rp_decorator[$setting_key]', function(value) {
                        value.bind(function(newval) {
                            newval = (newval !== '' ? newval : $original);
                            newval = prepare(newval, '$setting_key', '$selector');
                            jQuery('$selector').html(newval);
                        });
                    });";
                }
            }
        }

        // Close container and return
        echo $scripts . '});</script>';
    }




}

RP_Decorator_Preview::get_instance();

}


decorator.php
<?php

/**
 * Plugin Name: Decorator - WooCommerce Email Customizer
 * Plugin URI: http://www.rightpress.net/decorator
 * Description: Use native WordPress Customizer to make WooCommerce emails match your brand
 * Version: 1.0.2
 * Author: RightPress
 * Author URI: http://www.rightpress.net
 * Requires at least: 4.4
 * Tested up to: 4.7
 *
 * Text Domain: rp_decorator
 * Domain Path: /languages
 *
 * Copyright 2016 RightPress (email: info@rightpress.net)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @package Decorator
 * @category Core
 * @author RightPress
 */

// Exit if accessed directly
if (!defined('ABSPATH')) {
    exit;
}

// Define Constants
define('RP_DECORATOR_PLUGIN_PATH', plugin_dir_path(__FILE__));
define('RP_DECORATOR_PLUGIN_URL', plugins_url(basename(plugin_dir_path(__FILE__)), basename(__FILE__)));
define('RP_DECORATOR_VERSION', '1.0.2');
define('RP_DECORATOR_SUPPORT_PHP', '5.3');
define('RP_DECORATOR_SUPPORT_WP', '4.4');
define('RP_DECORATOR_SUPPORT_WC', '2.4');

if (!class_exists('RP_Decorator')) {

/**
 * Main plugin class
 *
 * @package Decorator
 * @author RightPress
 */
class RP_Decorator
{
    // Properties
    private static $admin_capability = null;

    // Singleton instance
    private static $instance = false;

    /**
     * Singleton control
     */
    public static function get_instance()
    {
        if (!self::$instance) {
            self::$instance = new self();
        }
        return self::$instance;
    }

    /**
     * Class constructor
     *
     * @access public
     * @return void
     */
    public function __construct()
    {
        // Load translation
        load_textdomain('rp_decorator', WP_LANG_DIR . '/decorator-woocommerce-email-customizer/rp_decorator-' . apply_filters('plugin_locale', get_locale(), 'rp_decorator') . '.mo');
        load_plugin_textdomain('rp_decorator', false, dirname(plugin_basename(__FILE__)) . '/languages/');

        // Execute other code when all plugins are loaded
        add_action('plugins_loaded', array($this, 'on_plugins_loaded'), 1);
    }

    /**
     * Code executed when all plugins are loaded
     *
     * @access public
     * @return void
     */
    public function on_plugins_loaded()
    {
        // Check environment
        if (!RP_Decorator::check_environment()) {
            return;
        }

        // Includes
        foreach (glob(RP_DECORATOR_PLUGIN_PATH . 'includes/classes/*.class.php') as $filename) {
            include $filename;
        }

        // Plugins page links
        add_filter('plugin_action_links_' . plugin_basename(__FILE__), array($this, 'plugins_page_links'));
    }

    /**
     * Check if environment meets requirements
     *
     * @access public
     * @return bool
     */
    public static function check_environment()
    {
        $is_ok = true;

        // Check PHP version
        if (!RP_Decorator::php_version_gte(RP_DECORATOR_SUPPORT_PHP)) {
            add_action('admin_notices', array('RP_Decorator', 'php_version_notice'));
        }

        // Check WordPress version
        if (!RP_Decorator::wp_version_gte(RP_DECORATOR_SUPPORT_WP)) {
            add_action('admin_notices', array('RP_Decorator', 'wp_version_notice'));
            $is_ok = false;
        }

        // Check if WooCommerce is enabled
        if (!class_exists('WooCommerce')) {
            add_action('admin_notices', array('RP_Decorator', 'wc_disabled_notice'));
            $is_ok = false;
        }
        else if (!RP_Decorator::wc_version_gte(RP_DECORATOR_SUPPORT_WC)) {
            add_action('admin_notices', array('RP_Decorator', 'wc_version_notice'));
            $is_ok = false;
        }

        return $is_ok;
    }

    /**
     * Check PHP version
     *
     * @access public
     * @param string $version
     * @return bool
     */
    public static function php_version_gte($version)
    {
        return version_compare(PHP_VERSION, $version, '>=');
    }

    /**
     * Check WordPress version
     *
     * @access public
     * @param string $version
     * @return bool
     */
    public static function wp_version_gte($version)
    {
        $wp_version = get_bloginfo('version');

        // Treat release candidate strings
        $wp_version = preg_replace('/-RC.+/i', '', $wp_version);

        if ($wp_version) {
            return version_compare($wp_version, $version, '>=');
        }

        return false;
    }

    /**
     * Check WooCommerce version
     *
     * @access public
     * @param string $version
     * @return bool
     */
    public static function wc_version_gte($version)
    {
        if (defined('WC_VERSION') && WC_VERSION) {
            return version_compare(WC_VERSION, $version, '>=');
        }
        else if (defined('WOOCOMMERCE_VERSION') && WOOCOMMERCE_VERSION) {
            return version_compare(WOOCOMMERCE_VERSION, $version, '>=');
        }
        else {
            return false;
        }
    }

    /**
     * Display PHP version notice
     *
     * @access public
     * @return void
     */
    public static function php_version_notice()
    {
        echo '<div class="error"><p>' . sprintf(__('<strong>Decorator</strong> requires PHP %s or later. Please update PHP on your server to use this plugin.', 'rp_decorator'), RP_DECORATOR_SUPPORT_PHP) . '</p></div>';
    }

    /**
     * Display WP version notice
     *
     * @access public
     * @return void
     */
    public static function wp_version_notice()
    {
        echo '<div class="error"><p>' . sprintf(__('<strong>Decorator</strong> requires WordPress version %s or later. Please update WordPress to use this plugin.', 'rp_decorator'), RP_DECORATOR_SUPPORT_WP) . '</p></div>';
    }

    /**
     * Display WC disabled notice
     *
     * @access public
     * @return void
     */
    public static function wc_disabled_notice()
    {
        echo '<div class="error"><p>' . sprintf(__('<strong>Decorator</strong> requires WooCommerce to be active. You can download WooCommerce %s.', 'rp_decorator'), '<a href="http://url.rightpress.net/woocommerce-download-page">' . __('here', 'rp_decorator') . '</a>') . '</p></div>';
    }

    /**
     * Display WC version notice
     *
     * @access public
     * @return void
     */
    public static function wc_version_notice()
    {
        echo '<div class="error"><p>' . sprintf(__('<strong>Decorator</strong> requires WooCommerce version %s or later. Please update WooCommerce to use this plugin.', 'rp_decorator'), RP_DECORATOR_SUPPORT_WC) . '</p></div>';
    }

    /**
     * Plugins page links
     *
     * @access public
     * @param array $links
     * @return array
     */
    public function plugins_page_links($links)
    {
        // Open Decorator
        $settings_link = '<a href="' . RP_Decorator_Customizer::get_customizer_url() . '">'.__('Open Decorator', 'rp_decorator').'</a>';
        array_unshift($links, $settings_link);

        return $links;
    }

    /**
     * Check if current user has administrative capability
     *
     * @access public
     * @return bool
     */
    public static function is_admin()
    {
        return current_user_can(RP_Decorator::get_admin_capability());
    }

    /**
     * Get admininistrative capability
     *
     * @access public
     * @return string
     */
    public static function get_admin_capability()
    {
        // Get capability
        if (self::$admin_capability === null) {
            self::$admin_capability = apply_filters('rp_decorator_capability', 'manage_woocommerce');
        }

        // Return capability
        return self::$admin_capability;
    }

    /**
     * Check if current request is own Customizer request
     *
     * @access public
     * @return bool
     */
    public static function is_own_customizer_request()
    {
        return isset($_REQUEST['rp-decorator-customize']) && $_REQUEST['rp-decorator-customize'] === '1';
    }

    /**
     * Check if current request is preview request
     *
     * @access public
     * @return bool
     */
    public static function is_own_preview_request()
    {
        return isset($_REQUEST['rp-decorator-preview']) && $_REQUEST['rp-decorator-preview'] === '1';
    }



}

RP_Decorator::get_instance();

}
